src/consensus/vote_tracker.zig

```diff
@@ -0,0 +1,343 @@
+    /// before releasing the guard on this lock (but not after).
+    ///
+    /// TODO: document whether a read guard on this map lock permits acquiring
+    /// a write guard on the entry's lock lock, and other potential restrictions
+    /// and/or allowances around safely accessing any data.
+    map_rwlock: std.Thread.RwLock,
+
+    /// Map from a slot to a set of validators who have voted for that slot.
+    /// See the doc comment on `map_rwlock` for commentary on accessing this field and its contents.
+    map: Map,
+
+    pub const EMPTY: VoteTracker = .{
+        .map_rwlock = .{},
+        .map = .{},
+    };
+
+    pub const Map = std.AutoArrayHashMapUnmanaged(Slot, *RcRwSlotVoteTracker);
+    pub const RcRwSlotVoteTracker = struct {
+        /// Must call `rc.acquire()` before sharing a reference to this slot vote tracker
+        /// with another thread. Should be done by any functions that directly get access
+        /// to a SVT, so that callers do not need to; a holder of a SVT needs to manage
+        /// the reference count appropriately after acquiring a SVT reference.
+        rc: sig.sync.ReferenceCounter,
+        tracker: sig.sync.RwMux(SlotVoteTracker),
+
+        pub fn create(allocator: std.mem.Allocator) std.mem.Allocator.Error!*RcRwSlotVoteTracker {
+            const self = try allocator.create(RcRwSlotVoteTracker);
+            self.* = .{
+                .rc = .{},
+                .tracker = sig.sync.RwMux(SlotVoteTracker).init(SlotVoteTracker.EMPTY_ZEROES),
+            };
+            return self;
+        }
+
+        /// Destroys `self` after freeing all of its resources if it is the last reference.
+        pub fn deinit(self: *RcRwSlotVoteTracker, allocator: std.mem.Allocator) void {
+            if (!self.rc.release()) return;
+            // can't unlock the rwmux, which is part of the freed memory,
+            // and in theory this is the only remaining reference anyway.
+            const tracker, _ = self.tracker.writeWithLock();
+            tracker.deinit(allocator);
+            allocator.destroy(self);
+        }
+    };
+
+    pub fn deinit(self: *VoteTracker, allocator: std.mem.Allocator) void {
+        self.map_rwlock.lock();
+        defer self.map_rwlock.unlock();
+        const map = &self.map;
+        for (map.values()) |rc_rw_svt| {
+            rc_rw_svt.deinit(allocator);
+        }
+        map.deinit(allocator);
+    }
+
+    /// The caller is responsible for calling `.deinit(allocator)` on the result.
+    pub fn getOrInsertSlotTracker(
+        self: *VoteTracker,
+        allocator: std.mem.Allocator,
+        slot: Slot,
+    ) std.mem.Allocator.Error!*RcRwSlotVoteTracker {
+        blk: {
+            self.map_rwlock.lockShared();
+            defer self.map_rwlock.unlockShared();
+            const rc_rw_svt = self.map.get(slot) orelse break :blk;
+            // we acquired a lock on the map, this shouldn't be possible
```

Comment:
Looks like this comment is not correct? Given we retrieved the `rc_rw_svt` and then we call rc_rw_svt.rc.acquire().

Reply:
To be clear, the intended reading of the comment is that the assertion shouldn't be triggered. Is that what you're saying is incorrect, or is the comment badly phrased?

Reply:
I think it's safe to assume that it will always return true here. If it returns false, that means there's another function that's destroying the item without removing it from the map, or another function is not respecting the lock. Either of those sound like bugs that assertions like this are supposed to help us identify.

In general, for typical reference counting use cases (where the counter itself gets destroyed with the resource), the return value of this function should never be false. If it's false, that means you have already triggered undefined behavior. You're actually lucky if it returns false instead of segfaulting or something worse.

Reply:
Well, there's also the possibility that `rc.release()` is called one too many times, without actually freeing any memory, which isn't that far-fetched.

Reply:
good point

Reply:
> or is the comment badly phrased?

I think having something like "this assert should not fail" instead of "this shouldn't be possible" may be more clearer?

Reply:
b8f0cafe528832ab9e34c3f9a4824e80855702ee